atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java

```diff
@@ -0,0 +1,201 @@
+                                            .build())),
+                    ColumnValueDescription.forType(ValueType.VAR_LONG))),
+            ConflictHandler.IGNORE_ALL,
+            TableMetadataPersistence.CachePriority.WARM,
+            false,
+            0,
+            false,
+            TableMetadataPersistence.SweepStrategy.NOTHING, // we do our own cleanup
+            false,
+            TableMetadataPersistence.LogSafety.SAFE);
+    private static final long WRITE_TIMESTAMP = 0L;
+    private static final ObjectMapper OBJECT_MAPPER = ObjectMappers.newServerObjectMapper();
+
+    private final KeyValueService kvs;
+    private final byte[] coordinationSequence;
+
+    private KeyValueServiceCoordinationStore(KeyValueService kvs, byte[] coordinationSequence) {
+        this.kvs = kvs;
+        this.coordinationSequence = coordinationSequence;
+    }
+
+    public static CoordinationStore create(KeyValueService kvs, byte[] coordinationSequence) {
+        CoordinationStore coordinationStore = new KeyValueServiceCoordinationStore(kvs, coordinationSequence);
+        kvs.createTable(AtlasDbConstants.COORDINATION_TABLE, COORDINATION_TABLE_METADATA.persistToBytes());
+        return coordinationStore;
+    }
+
+    @Override
+    public Optional<ByteString> getValue(long sequenceNumber) {
+        Preconditions.checkState(
+                sequenceNumber > 0,
+                "Only positive sequence numbers are supported, but found %s",
+                sequenceNumber);
+        Cell targetCell = getCellForSequence(sequenceNumber);
+        Map<Cell, Value> response = kvs.get(AtlasDbConstants.COORDINATION_TABLE,
+                ImmutableMap.of(targetCell, Long.MAX_VALUE));
+        return Optional.ofNullable(response.get(targetCell)).map(Value::getContents).map(ByteString::of);
+    }
+
+    @Override
+    public void putValue(long sequenceNumber, ByteString value) {
+        Preconditions.checkState(
+                sequenceNumber > 0,
+                "Only positive sequence numbers are supported, but found %s",
+                sequenceNumber);
+        kvs.put(AtlasDbConstants.COORDINATION_TABLE,
+                ImmutableMap.of(getCellForSequence(sequenceNumber), value.toByteArray()),
+                WRITE_TIMESTAMP);
+    }
+
+    @Override
+    public Optional<SequenceAndBound> getCoordinationValue() {
+        Cell targetCell = getCoordinationValueCell();
+        Map<Cell, Value> response
+                = kvs.get(AtlasDbConstants.COORDINATION_TABLE, ImmutableMap.of(targetCell, Long.MAX_VALUE));
+        return Optional.ofNullable(response.get(targetCell))
+                .map(Value::getContents)
+                .map(KeyValueServiceCoordinationStore::deserializeSequenceAndBound);
+    }
+
+    @Override
+    public CheckAndSetResult<SequenceAndBound> checkAndSetCoordinationValue(
+            Optional<SequenceAndBound> oldValue, SequenceAndBound newValue) {
+        CheckAndSetRequest request = oldValue
```

Comment:
I'm fine either way, but could be easier to read using CheckAndSetRequest.Builder directly

Reply:
That is _much_ better actually, it becomes

```
        CheckAndSetRequest request = new CheckAndSetRequest.Builder()
                .table(AtlasDbConstants.COORDINATION_TABLE)
                .cell(getCoordinationValueCell())
                .oldValue(oldValue.map(KeyValueServiceCoordinationStore::serializeSequenceAndBound))
                .newValue(serializeSequenceAndBound(newValue))
                .build();
```